src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,139 @@
+            // Can only ensure a break scoped in a block is deterministic
+            continue;
+        } else if (containsBreakOrReturn(s.getChildren())) {
+            return true;
+        }
     }
 
+    return false;
+};
+
+export const transformSwitchStatement: FunctionVisitor<ts.SwitchStatement> = (statement, context) => {
     const scope = pushScope(context, ScopeType.Switch);
 
-    // Give the switch a unique name to prevent nested switches from acting up.
+    // Give the switch and condition accumulator a unique name to prevent nested switches from acting up.
     const switchName = `____switch${scope.id}`;
+    const conditionName = `____cond${scope.id}`;
     const switchVariable = lua.createIdentifier(switchName);
+    const conditionVariable = lua.createIdentifier(conditionName);
 
+    // If the switch only has a default clause, wrap it in a single do.
+    // Otherwise, we need to generate a set of if statements to emulate the switch.
     let statements: lua.Statement[] = [];
-
-    // Starting from the back, concatenating ifs into one big if/elseif statement
-    const concatenatedIf = statement.caseBlock.clauses.reduceRight((previousCondition, clause, index) => {
-        if (ts.isDefaultClause(clause)) {
-            // Skip default clause here (needs to be included to ensure index lines up with index later)
-            return previousCondition;
+    const clauses = statement.caseBlock.clauses;
+    if (clauses.length === 1 && ts.isDefaultClause(clauses[0])) {
+        const defaultClause = clauses[0].statements;
+        if (defaultClause.length) {
+            statements.push(lua.createDoStatement(context.transformStatements(defaultClause)));
         }
+    } else {
+        // Build up the condition for each if statement
+        let isInitialCondition = true;
+        let condition: lua.Expression | undefined = undefined;
+        for (let i = 0; i < statement.caseBlock.clauses.length; i++) {
+            const clause = statement.caseBlock.clauses[i];
+            const previousClause: ts.CaseOrDefaultClause | undefined = statement.caseBlock.clauses[i - 1];
 
-        // If the clause condition holds, go to the correct label
-        const condition = lua.createBinaryExpression(
-            switchVariable,
-            context.transformExpression(clause.expression),
-            lua.SyntaxKind.EqualityOperator
-        );
+            // Skip redundant default clauses, will be handled in final default case
+            if (i === 0 && ts.isDefaultClause(clause)) continue;
+            if (ts.isDefaultClause(clause) && previousClause && containsBreakOrReturn([...previousClause.statements])) {
+                continue;
+            }
 
-        const goto = lua.createGotoStatement(`${switchName}_case_${index}`);
-        return lua.createIfStatement(condition, lua.createBlock([goto]), previousCondition);
-    }, undefined as lua.IfStatement | undefined);
+            // Compute the condition for the if statement
```

Comment:
Candidate for refactoring to its own function?

Reply:
Sure! Could be easier on the eyes here.